src/ir_passes/lower/fsm.rs

```diff
@@ -1,17 +1,406 @@
+                let this = builder.component.signature.borrow();
+
+                for fsm in 0..fsm_num {
+                    let c = counters[fsm as usize].borrow();
+
+                    let go = if fsm == 0 {
+                        // this is the first fsm, start it when go is high
+                        this.get("go")
+                    } else {
+                        // start this fsm when the previous one is done.
+                        counters[(fsm - 1) as usize].borrow().get("done")
+                    };
+
+                    // hook up the end of the last fsm to this one's start.
+                    builder.component.continuous_assignments.push(
+                        builder.build_assignment(
+                            c.get("go"),
+                            go,
+                            calyx::Guard::<Nothing>::True,
+                        ),
+                    );
+
+                    // build assignments to state output ports
+                    for st in 0..delay {
+                        builder.component.continuous_assignments.push(
+                            builder.build_assignment(
+                                this.get(format!("_{}", fsm * delay + st)),
+                                c.get(format!("_{st}")),
+                                calyx::Guard::<Nothing>::True,
+                            ),
+                        );
+                    }
+                }
+
+                // done <= _{n-1};
+                // unused at the moment but useful if we want to chain FSMs.
+                builder.component.continuous_assignments.push(
+                    builder.build_assignment(
+                        this.get("done"),
+                        counters[(fsm_num - 1) as usize].borrow().get("done"),
+                        Guard::True,
+                    ),
+                );
+
+                drop(this);
+                comp
+            })
+    }
+
+    /// Helper function that generates a [calyx:Component] for a counter with `n` states.
+    fn add_counter(
+        &'l mut self,
+        states: u64,
+        lib: Option<&'l calyx::LibrarySignatures>,
+    ) -> &'l calyx::Component {
+        self.counter.entry(states).or_insert_with(|| {
+            // gets the number of bits needed to represent the counter state.
+            let bitwidth = (64 - (states - 1).leading_zeros()) as u64;
+
+            let mut comp = FsmBind::fsm_comp("counter", states);
+            let mut builder =
+                calyx::Builder::new(&mut comp, lib.unwrap()).not_generated();
+
+            let state = builder.add_primitive("state", "std_reg", &[bitwidth]);
+
+            let add = builder.add_primitive("add", "std_add", &[bitwidth]);
+
+            // Constant signal
+            structure!(builder;
+                let signal_on = constant(1, 1);
+                let one = constant(1, bitwidth);
+                let zero = constant(0, bitwidth);
+            );
```

Comment:
Oops, I see you already use them! In that case, the above should be folded into this `structure!` call.